Rewrite transaction docs for address balances and tx.coin/tx.balance (#897)
MystenLabs/ts-sdks: packages/docs/content/sui/transactions/coins-and-balances.mdx

---
title: Coins and Balances
description: Work with coin objects and address balances in transactions
---

> Sui has two systems for holding fungible token balances: **coin objects** and **address balances**.

Reviewer: ```suggestion
Sui has two systems for holding fungible token balances: coin objects and address balances.
```

> **Coin objects** are individual on-chain objects, each with its own ID, version, and balance. You

Reviewer: ```suggestion
**Coin objects** are individual onchain objects, each with its own ID, version, and balance. You
```

own specific coin objects and need to split, merge, and track them.

> **Address balances** are an accumulator per address per coin type. There are no objects to manage —

Reviewer: ```suggestion
**Address balances** are an accumulator per address per coin type. There are no objects to manage because
```

deposits automatically merge into a single balance, and you withdraw from it as needed.

Both systems coexist. An address's total balance for a given coin type is the sum of its coin object
balances and its address balance.

Address balances have no object versions. When a transaction has no versioned object inputs (for
example, a balance transfer built entirely from address balance withdrawals), it won't be
invalidated by other transactions from the same address executing concurrently.

> ## tx.coin and tx.balance

Reviewer: ```suggestion
## `tx.coin` and `tx.balance`
```

`tx.coin()` and `tx.balance()` are the recommended ways to get tokens in a transaction. They
automatically draw from both coin objects and address balances.

> ### Getting a Coin

Reviewer: ```suggestion
### Getting a `Coin`
```

> `tx.coin()` produces a `Coin<T>` — use it for transfers and most operations:

Reviewer: ```suggestion
`tx.coin()` produces a `Coin<T>`. Use it for transfers and most operations:
```

```tsx
import { Transaction } from '@mysten/sui/transactions';

const tx = new Transaction();

// SUI (balance is in MIST — 1 SUI = 1,000,000,000 MIST)
tx.transferObjects([tx.coin({ balance: 1_000_000_000n })], '0xRecipientAddress');

// Another coin type
tx.transferObjects(
	[tx.coin({ balance: 1_000_000n, type: '0xPackageId::module::CoinType' })],
	'0xRecipientAddress',
);
```

> ### Getting a Balance

Reviewer: ```suggestion
### Getting a `Balance`
```

> `tx.balance()` produces a `Balance<T>` — use it for Move functions that expect a balance directly,

Reviewer: ```suggestion
`tx.balance()` produces a `Balance<T>`. Use it for Move functions that expect a balance directly,
```

	arguments: [tx.balance({ balance: 1_000_000_000n }), tx.pure.address('0xRecipientAddress')],
});
```

<Callout type="info">
	Transactions built entirely from `tx.balance()` and gasless-eligible Move calls like `send_funds`
	and `redeem_funds` may qualify as [gasless transactions](#gasless-transactions).
</Callout>

### Options

| Option       | Type               | Default | Description                                                              |
| ------------ | ------------------ | ------- | ------------------------------------------------------------------------ |
| `balance`    | `bigint \| number` | —       | Amount in base units (MIST for SUI)                                      |
| `type`       | `string`           | SUI     | Coin type. Defaults to `0x2::sui::SUI`                                   |
| `useGasCoin` | `boolean`          | `true`  | For SUI, split from the gas coin. Set `false` for sponsored transactions |

For SUI, `tx.coin()` splits from the gas coin by default. For sponsored transactions where the gas
coin belongs to the sponsor, set `useGasCoin: false`:

```tsx
tx.transferObjects([tx.coin({ balance: 100n, useGasCoin: false })], recipient);
```

> ### coinWithBalance

Reviewer: ```suggestion
### `coinWithBalance`
```

`coinWithBalance()` is a standalone alias for `tx.coin()`:

```tsx
import { coinWithBalance, Transaction } from '@mysten/sui/transactions';

const tx = new Transaction();
tx.transferObjects([coinWithBalance({ balance: 1_000_000_000 })], recipient);
```

## How resolution works

When you call `tx.coin()` or `tx.balance()`, the SDK adds a placeholder intent to the transaction.
At build time, the resolver replaces it with concrete commands based on the sender's funds:

- **`tx.balance()` with sufficient address balance:** Uses a direct `FundsWithdrawal` via
  `balance::redeem_funds`. No coin objects are used, so the transaction has no versioned object

>   inputs from this intent — keeping it eligible for parallel execution.

Reviewer: ```suggestion
  inputs from this intent, keeping it eligible for parallel execution.
```

> - **Otherwise (coins needed):** Fetches the sender's coin objects and address balance in parallel.

Reviewer: ```suggestion
- **Otherwise, coins are needed:** Fetches the sender's coin objects and address balance in parallel.
```
